File: 序号/1870. 准时到达的列车最小时速.py
class Solution:
    def minSpeedOnTime(self, dist: List[int], hour: float) -> int:
        if hour <= len(dist)-1:
            return -1
        l,r=0,int(1e7+1)
        while l+1<r:
            m=(l+r)//2
            sum=0
            for i in range(len(dist)-1):# 最后一班车不用向上取整
                sum+=ceil(dist[i]/m)
            sum+=dist[-1]/m
            if sum<=hour:
                r=m
            else:
                l=m
        return r

File: 序号/1870. 准时到达的列车最小时速.kt
class Solution {
    fun minSpeedOnTime(dist: IntArray, hour: Double): Int {
        val busCnt = dist.size
        if(hour<=busCnt-1) return -1 // 等于就要求最后辆车速度无限快
        var l=0
        var r = Int.MAX_VALUE
        fun check(v:Int):Boolean{
            var t = 0
            for(i in 0 until dist.size-1){
                val d = dist[i]
                if(d%v==0) t+=d/v
                else t+=d/v+1
            }
            return t+dist.last().toDouble()/v<=hour
        }
        while(l+1<r){
            val m = l+(r-l)/2
            if(check(m)){
                r=m
            }else{
                l=m
            }
        }
        return r
    }
}